### DISCUSSION TITLE
Implementing both communication over wifi (from AIdeck) and over radio (to-from code in app layer)

### DISCUSSION THREAD

**Discussion by matsarto99** [bitcraze/discussions#719] [Q&A] [answered]
Hello, 

For what I am trying to do I would need the AI deck to stream images over wifi, collect and process those images on my PC and then send back some info to the drone which would perform a navigation task.

For the first step, I am using the simple wifi-img-streamer demo with the wifi module set as access point. In the second part, I am connecting to the drone via crazyradio and sending data to the app layer in a way similar to what the app_appchannel_test does. The third step would indeed be implemented as an app in the app layer.

The first question that I would like to ask is if this setting does make sense or if it would be easier to send data back over wifi using CPX and then from the GAP8 to the STM32 via UART. I thought that setting up the communication via radio and using CRTP could be faster and would communicate "directly" to the app layer, but I might be wrong.

I could maybe have answered the question myself trying, but here comes my main issue.. I am not able to flash the code for the app layer and keep the wifi as access point at the same time. To explain better:
- I set up the access point using `make menuconfig` from the crazyflie-firmware folder
- I flash the AI deck with the wifi-img-streamer demo and it works
- I use the toolbelt to make the app code using `tb make_app path/to/app`

Now:
- If I flash from the app directory, the settings that keep the wifi as access point disappear and i cannot receive the stream anymore
- If I flash from the firmware directory, the app layer configuration is not set (not enabled from what I can see at build/.config)

To flash I am using `cfloader -w [crazyflie_uri] flash build/cf2.bin stm32-fw` .

Am I doing something wrong? How can I flash the firmware with the app layer code and still keep the config of the wifi access point?

I found something similar in #151 and #181 but I am not really sure updating the ESP firmware is the problem here.
Edit: found #335, but if I disable the wifi AP from menuconfig I cannot see the AP anymore. The image streamer is setting up a connection, but not configuring the AP from what I got.. Indeed enabling the AP from menuconfig is a step which is required in the image streamer according to the documentation.

Thanks in advance

Mattia

**Comment by ataffanel** [comment]
Hi,

First of all I think your strategy is good: Crazyradio will indeed allow you to have low latency direct communication with apps and the Crazyflie.

I think, what might be happening, is that you are not enabling the wifi image streamer in your app: when making an app for the Crazyflie, your app folder becomes a new full project. One way to think about it is that your app is using the Crazyflie firmware *as a lib*. So you likely need to `make menuconfig` from within your app folder.

**Reply by matsarto99** [reply]
Hello,

Thanks for your answer!
I tried to run `make menuconfig` from within the app folder, but I get:
make: *** No rule to make target 'menuconfig'.  Stop.

I tried also in one of the already provided examples, but I get the same message.

Any hints? Do I have to set up something manually?

Mattia

**Reply by krichardsson** [reply, marked as answer]
I don't think menuconfig works in an app folder, but there is a file called `app-config` that you can edit manually. 
You can run menuconfig in the root directory and check the output in the `build/.config` to get the proper names of the config settings

**Reply by matsarto99** [reply]
I tried putting the configurations for the access point in the `.app-config` file and now I can see the AP active. Specifically I added:
```
CONFIG_DECK_AI_WIFI_SETUP_AP=y
CONFIG_DECK_AI_SSID="myssid"
CONFIG_DECK_AI_PASSWORD="myssidpwd"
```
I am now getting issues communicating over the crazyradio, but I'll try to figure out what is happening and update you

Edit: Thanks Kristoffer, I didn't see your answer, but I think I ended up doing what you suggested

**Reply by matsarto99** [reply]
Hello,

After some time, I think I also understood the problem with the Crazyradio. Every time I switch the wifi connection the radio gets detached from the WSL. Moreover, if the code is supposed to use the radio and the radio is suddenly not reachable (in my case because of the change in wifi connection), then the WSL gets stuck and I have to restart it.

As an example: I was connected to the drone access point when I started flashing a new firmware -> the drone switched to bootloader mode and the access point got disabled -> my PC connected to another available network -> the CrazyRadio got detached from the WSL -> the flashing got stuck uncompleted -> since it was using the radio and this was suddenly detached, the WSL got stuck (maybe it was busy trying to reach a resource that was no more available).

Right now my solution is to manually detach the radio on Windows side via the commands provided with usbipd-win and then attach it again every time I have to switch wifi network (to-from drone AP). I am not sure whether it is a problem on CrazyRadio side or rather WSL side...

I know this issue is slightly different, but I think it is still relevant in this thread if someone wants to implement both wifi and radio communication. Anyway, let me know if I should start another discussion.

**Reply by knmcguire** [reply]
hi! This is not something we have used ourselves with WSL before, and I'm afraid we will not be able to get the solution how to fix it either... Could you try it on a native ubuntu computer or perhaps the VM?

**Reply by matsarto99** [reply]
Hello,
After some time I am back, I did not have a native Ubuntu computer with which to try. Now I do, I tried again and the issue seems to be solved so I guess it is a problem with WSL or my PC.

**Reply by krichardsson** [reply]
Thanks for the update!
